util/visualization/gemmx.py

```diff
@@ -0,0 +1,275 @@
+
+    # Create bars for stalls and fires
+    ax.bar(indices, number_of_stalls, bar_width, label="Number of Stalls")
+    ax.bar(indices + bar_width, number_of_fires, bar_width, label="Number of Fires")
+
+    # Labeling and aesthetics
+    ax.set_xlabel("Port Number", fontsize=14)
+    ax.set_ylabel("Count", fontsize=14)
+    ax.set_title("Number of Stalls and Fires per Port of GeMMX", fontsize=16)
+    ax.set_xticks(indices + bar_width / 2)
+    ax.set_xticklabels(ports, rotation=90)
+    ax.legend()
+
+    # Annotate port groups
+    group_positions = [
+        (0, 7, "A"),  # Ports 0-7
+        (8, 15, "B"),  # Ports 8-15
+        (16, 23, "D8"),  # Ports 16-23
+        (24, 55, "C/D32"),  # Ports 24-55
+    ]
+
+    # Add annotations
+    for start, end, label in group_positions:
+        center = (indices[start] + indices[end] + bar_width) / 2
+        ax.hlines(
+            -8, indices[start], indices[end] + bar_width, color="black", linewidth=2
+        )
+        ax.text(center, -11, label, ha="center", va="top", fontsize=12)
+
+    # Adjust plot limits
+    ax.set_ylim(bottom=-20)
+    ax.spines["bottom"].set_position(("data", 0))
+    plt.tight_layout()
+
+    # Save the plot
+    output_file = os.path.join(output_path, "nb_of_stalls_per_port.png")
+    plt.savefig(output_file)
+    plt.close()
+
+
+def plot_stalls_and_fires_per_bank(df, output_path):
+    """
+    Creates a bar plot of the number of stalls and fires per bank.
+    """
+    banks = np.arange(32)
+    number_of_stalls = np.zeros(32)
+    number_of_fires = np.zeros(32)
+
+    for bank_nb in range(32):
+        nb_stalls = 0
+        nb_fires = 0
+        for port in range(56):
+            stall_col = f"io_data_tcdm_req_{port}_stall"
+            fire_col = f"io_data_tcdm_req_{port}_fire"
+            bank_col = f"io_data_tcdm_req_{port}_bank"
+
+            stalls = df[(df[stall_col]) & (df[bank_col] == bank_nb)]
+            fires = df[(df[fire_col]) & (df[bank_col] == bank_nb)]
+
+            nb_stalls += stalls.shape[0]
+            nb_fires += fires.shape[0]
+
+        number_of_stalls[bank_nb] = nb_stalls
+        number_of_fires[bank_nb] = nb_fires
+
+    # Plotting
+    fig, ax = plt.subplots(figsize=(18, 8))
+    bar_width = 0.35
+    indices = np.arange(len(banks))
+
+    # Create bars for stalls and fires
+    ax.bar(indices, number_of_stalls, bar_width, label="Number of Stalls")
+    ax.bar(indices + bar_width, number_of_fires, bar_width, label="Number of Fires")
+
+    # Labeling and aesthetics
+    ax.set_xlabel("Bank Number", fontsize=14)
+    ax.set_ylabel("Count", fontsize=14)
+    ax.set_title("Number of Stalls and Fires per Bank of TCDM", fontsize=16)
+    ax.set_xticks(indices + bar_width / 2)
+    ax.set_xticklabels(banks, rotation=90)
+    ax.legend()
+
+    plt.tight_layout()
+
+    # Save the plot
+    output_file = os.path.join(output_path, "nb_of_stalls_per_bank.png")
+    plt.savefig(output_file)
+    plt.close()
```

Comment:
maybe put this in `main()`?

Reply:
no because I may want to use the functions separately